File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/ahn-product-1-0/EventCheckBase.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class EventCheckBase : IEventCheck
    {
        public EventCheckBase(XmlNode xmlNode)
        : base(xmlNode)
        {
            XmlNode eventIndexTransformationNode = xmlNode.SelectSingleNode("eventIndexTransformation");
            
            if (eventIndexTransformationNode != null)
            {
                if (eventIndexTransformationNode.Attributes["href"] != null || eventIndexTransformationNode.Attributes["id"] != null) 
                {
                    if (eventIndexTransformationNode.Attributes["id"] != null) 
                    {
                        eventIndexTransformationIDRef_ = eventIndexTransformationNode.Attributes["id"].Value;
                        EventIndexTransformation ob = new EventIndexTransformation(eventIndexTransformationNode);
                        IDManager.SetID(eventIndexTransformationIDRef_, ob);
                    }
                    else if (eventIndexTransformationNode.Attributes["href"] != null)
                    {
                        eventIndexTransformationIDRef_ = eventIndexTransformationNode.Attributes["href"].Value;
                    }
                    else
                    {
                        eventIndexTransformation_ = new EventIndexTransformation(eventIndexTransformationNode);
                    }
                }
                else
                {
                    eventIndexTransformation_ = new EventIndexTransformation(eventIndexTransformationNode);
                }
            }
            
        
            XmlNode functionEventsNode = xmlNode.SelectSingleNode("functionEvents");
            
            if (functionEventsNode != null)
            {
                if (functionEventsNode.Attributes["href"] != null || functionEventsNode.Attributes["id"] != null) 
                {
                    if (functionEventsNode.Attributes["id"] != null) 
                    {
                        functionEventsIDRef_ = functionEventsNode.Attributes["id"].Value;
                        FunctionEvents ob = new FunctionEvents(functionEventsNode);
                        IDManager.SetID(functionEventsIDRef_, ob);
                    }
                    else if (functionEventsNode.Attributes["href"] != null)
                    {
                        functionEventsIDRef_ = functionEventsNode.Attributes["href"].Value;
                    }
                    else
                    {
                        functionEvents_ = new FunctionEvents(functionEventsNode);
                    }
                }
                else
                {
                    functionEvents_ = new FunctionEvents(functionEventsNode);
                }
            }
            
        
            XmlNode payoffBaseNode = xmlNode.SelectSingleNode("payoffBase");
            
            if (payoffBaseNode != null)
            {
                if (payoffBaseNode.Attributes["href"] != null || payoffBaseNode.Attributes["id"] != null) 
                {
                    if (payoffBaseNode.Attributes["id"] != null) 
                    {
                        payoffBaseIDRef_ = payoffBaseNode.Attributes["id"].Value;
                        PayoffBase ob = new PayoffBase(payoffBaseNode);
                        IDManager.SetID(payoffBaseIDRef_, ob);
                    }
                    else if (payoffBaseNode.Attributes["href"] != null)
                    {
                        payoffBaseIDRef_ = payoffBaseNode.Attributes["href"].Value;
                    }
                    else
                    {
                        payoffBase_ = new PayoffBase(payoffBaseNode);
                    }
                }
                else
                {
                    payoffBase_ = new PayoffBase(payoffBaseNode);
                }
            }
            
        
            XmlNode subEventNode = xmlNode.SelectSingleNode("subEvent");
            
            if (subEventNode != null)
            {
                if (subEventNode.Attributes["href"] != null || subEventNode.Attributes["id"] != null) 
                {
                    if (subEventNode.Attributes["id"] != null) 
                    {
                        subEventIDRef_ = subEventNode.Attributes["id"].Value;
                        SubEvent ob = new SubEvent(subEventNode);
                        IDManager.SetID(subEventIDRef_, ob);
                    }
                    else if (subEventNode.Attributes["href"] != null)
                    {
                        subEventIDRef_ = subEventNode.Attributes["href"].Value;
                    }
                    else
                    {
                        subEvent_ = new SubEvent(subEventNode);
                    }
                }
                else
                {
                    subEvent_ = new SubEvent(subEventNode);
                }
            }
            
        
            XmlNode jointSubEventNode = xmlNode.SelectSingleNode("jointSubEvent");
            
            if (jointSubEventNode != null)
            {
                if (jointSubEventNode.Attributes["href"] != null || jointSubEventNode.Attributes["id"] != null) 
                {
                    if (jointSubEventNode.Attributes["id"] != null) 
                    {
                        jointSubEventIDRef_ = jointSubEventNode.Attributes["id"].Value;
                        JointSubEvent ob = new JointSubEvent(jointSubEventNode);
                        IDManager.SetID(jointSubEventIDRef_, ob);
                    }
                    else if (jointSubEventNode.Attributes["href"] != null)
                    {
                        jointSubEventIDRef_ = jointSubEventNode.Attributes["href"].Value;
                    }
                    else
                    {
                        jointSubEvent_ = new JointSubEvent(jointSubEventNode);
                    }
                }
                else
                {
                    jointSubEvent_ = new JointSubEvent(jointSubEventNode);
                }
            }
            
        
            XmlNode jointFSubEventNode = xmlNode.SelectSingleNode("jointFSubEvent");
            
            if (jointFSubEventNode != null)
            {
                if (jointFSubEventNode.Attributes["href"] != null || jointFSubEventNode.Attributes["id"] != null) 
                {
                    if (jointFSubEventNode.Attributes["id"] != null) 
                    {
                        jointFSubEventIDRef_ = jointFSubEventNode.Attributes["id"].Value;
                        JointFSubEvent ob = new JointFSubEvent(jointFSubEventNode);
                        IDManager.SetID(jointFSubEventIDRef_, ob);
                    }
                    else if (jointFSubEventNode.Attributes["href"] != null)
                    {
                        jointFSubEventIDRef_ = jointFSubEventNode.Attributes["href"].Value;
                    }
                    else
                    {
                        jointFSubEvent_ = new JointFSubEvent(jointFSubEventNode);
                    }
                }
                else
                {
                    jointFSubEvent_ = new JointFSubEvent(jointFSubEventNode);
                }
            }
            
        
        }
        
    
        #region EventIndexTransformation_
        private EventIndexTransformation eventIndexTransformation_;
        public EventIndexTransformation EventIndexTransformation_
        {
            get
            {
                if (this.eventIndexTransformation_ != null)
                {
                    return this.eventIndexTransformation_; 
                }
                else if (this.eventIndexTransformationIDRef_ != null)
                {
                    eventIndexTransformation_ = IDManager.getID(eventIndexTransformationIDRef_) as EventIndexTransformation;
                    return this.eventIndexTransformation_; 
                }
                else
                {
                      return this.eventIndexTransformation_; 
                }
            }
            set
            {
                if (this.eventIndexTransformation_ != value)
                {
                    this.eventIndexTransformation_ = value;
                }
            }
        }
        #endregion
        
        public string eventIndexTransformationIDRef_ { get; set; }
        #region FunctionEvents_
        private FunctionEvents functionEvents_;
        public FunctionEvents FunctionEvents_
        {
            get
            {
                if (this.functionEvents_ != null)
                {
                    return this.functionEvents_; 
                }
                else if (this.functionEventsIDRef_ != null)
                {
                    functionEvents_ = IDManager.getID(functionEventsIDRef_) as FunctionEvents;
                    return this.functionEvents_; 
                }
                else
                {
                      return this.functionEvents_; 
                }
            }
            set
            {
                if (this.functionEvents_ != value)
                {
                    this.functionEvents_ = value;
                }
            }
        }
        #endregion
        
        public string functionEventsIDRef_ { get; set; }
        #region PayoffBase_
        private PayoffBase payoffBase_;
        public PayoffBase PayoffBase_
        {
            get
            {
                if (this.payoffBase_ != null)
                {
                    return this.payoffBase_; 
                }
                else if (this.payoffBaseIDRef_ != null)
                {
                    payoffBase_ = IDManager.getID(payoffBaseIDRef_) as PayoffBase;
                    return this.payoffBase_; 
                }
                else
                {
                      return this.payoffBase_; 
                }
            }
            set
            {
                if (this.payoffBase_ != value)
                {
                    this.payoffBase_ = value;
                }
            }
        }
        #endregion
        
        public string payoffBaseIDRef_ { get; set; }
        #region SubEvent_
        private SubEvent subEvent_;
        public SubEvent SubEvent_
        {
            get
            {
                if (this.subEvent_ != null)
                {
                    return this.subEvent_; 
                }
                else if (this.subEventIDRef_ != null)
                {
                    subEvent_ = IDManager.getID(subEventIDRef_) as SubEvent;
                    return this.subEvent_; 
                }
                else
                {
                      return this.subEvent_; 
                }
            }
            set
            {
                if (this.subEvent_ != value)
                {
                    this.subEvent_ = value;
                }
            }
        }
        #endregion
        
        public string subEventIDRef_ { get; set; }
        #region JointSubEvent_
        private JointSubEvent jointSubEvent_;
        public JointSubEvent JointSubEvent_
        {
            get
            {
                if (this.jointSubEvent_ != null)
                {
                    return this.jointSubEvent_; 
                }
                else if (this.jointSubEventIDRef_ != null)
                {
                    jointSubEvent_ = IDManager.getID(jointSubEventIDRef_) as JointSubEvent;
                    return this.jointSubEvent_; 
                }
                else
                {
                      return this.jointSubEvent_; 
                }
            }
            set
            {
                if (this.jointSubEvent_ != value)
                {
                    this.jointSubEvent_ = value;
                }
            }
        }
        #endregion
        
        public string jointSubEventIDRef_ { get; set; }
        #region JointFSubEvent_
        private JointFSubEvent jointFSubEvent_;
        public JointFSubEvent JointFSubEvent_
        {
            get
            {
                if (this.jointFSubEvent_ != null)
                {
                    return this.jointFSubEvent_; 
                }
                else if (this.jointFSubEventIDRef_ != null)
                {
                    jointFSubEvent_ = IDManager.getID(jointFSubEventIDRef_) as JointFSubEvent;
                    return this.jointFSubEvent_; 
                }
                else
                {
                      return this.jointFSubEvent_; 
                }
            }
            set
            {
                if (this.jointFSubEvent_ != value)
                {
                    this.jointFSubEvent_ = value;
                }
            }
        }
        #endregion
        
        public string jointFSubEventIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/ahn-product-1-0/EventCheckBase.cpp
// EventCheckBase.cpp 
#include "EventCheckBase.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

EventCheckBase::EventCheckBase(TiXmlNode* xmlNode)
: IEventCheck(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //eventIndexTransformationNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* eventIndexTransformationNode = xmlNode->FirstChildElement("eventIndexTransformation");

   if(eventIndexTransformationNode){eventIndexTransformationIsNull_ = false;}
   else{eventIndexTransformationIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- eventIndexTransformationNode , address : " << eventIndexTransformationNode << std::endl;
   #endif
   if(eventIndexTransformationNode)
   {
      if(eventIndexTransformationNode->Attribute("href") || eventIndexTransformationNode->Attribute("id"))
      {
          if(eventIndexTransformationNode->Attribute("id"))
          {
             eventIndexTransformationIDRef_ = eventIndexTransformationNode->Attribute("id");
             eventIndexTransformation_ = boost::shared_ptr<EventIndexTransformation>(new EventIndexTransformation(eventIndexTransformationNode));
             eventIndexTransformation_->setID(eventIndexTransformationIDRef_);
             IDManager::instance().setID(eventIndexTransformationIDRef_,eventIndexTransformation_);
          }
          else if(eventIndexTransformationNode->Attribute("href")) { eventIndexTransformationIDRef_ = eventIndexTransformationNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { eventIndexTransformation_ = boost::shared_ptr<EventIndexTransformation>(new EventIndexTransformation(eventIndexTransformationNode));}
   }

   //functionEventsNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* functionEventsNode = xmlNode->FirstChildElement("functionEvents");

   if(functionEventsNode){functionEventsIsNull_ = false;}
   else{functionEventsIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- functionEventsNode , address : " << functionEventsNode << std::endl;
   #endif
   if(functionEventsNode)
   {
      if(functionEventsNode->Attribute("href") || functionEventsNode->Attribute("id"))
      {
          if(functionEventsNode->Attribute("id"))
          {
             functionEventsIDRef_ = functionEventsNode->Attribute("id");
             functionEvents_ = boost::shared_ptr<FunctionEvents>(new FunctionEvents(functionEventsNode));
             functionEvents_->setID(functionEventsIDRef_);
             IDManager::instance().setID(functionEventsIDRef_,functionEvents_);
          }
          else if(functionEventsNode->Attribute("href")) { functionEventsIDRef_ = functionEventsNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { functionEvents_ = boost::shared_ptr<FunctionEvents>(new FunctionEvents(functionEventsNode));}
   }

   //payoffBaseNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* payoffBaseNode = xmlNode->FirstChildElement("payoffBase");

   if(payoffBaseNode){payoffBaseIsNull_ = false;}
   else{payoffBaseIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- payoffBaseNode , address : " << payoffBaseNode << std::endl;
   #endif
   if(payoffBaseNode)
   {
      if(payoffBaseNode->Attribute("href") || payoffBaseNode->Attribute("id"))
      {
          if(payoffBaseNode->Attribute("id"))
          {
             payoffBaseIDRef_ = payoffBaseNode->Attribute("id");
             payoffBase_ = boost::shared_ptr<PayoffBase>(new PayoffBase(payoffBaseNode));
             payoffBase_->setID(payoffBaseIDRef_);
             IDManager::instance().setID(payoffBaseIDRef_,payoffBase_);
          }
          else if(payoffBaseNode->Attribute("href")) { payoffBaseIDRef_ = payoffBaseNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { payoffBase_ = boost::shared_ptr<PayoffBase>(new PayoffBase(payoffBaseNode));}
   }

   //subEventNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* subEventNode = xmlNode->FirstChildElement("subEvent");

   if(subEventNode){subEventIsNull_ = false;}
   else{subEventIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- subEventNode , address : " << subEventNode << std::endl;
   #endif
   if(subEventNode)
   {
      if(subEventNode->Attribute("href") || subEventNode->Attribute("id"))
      {
          if(subEventNode->Attribute("id"))
          {
             subEventIDRef_ = subEventNode->Attribute("id");
             subEvent_ = boost::shared_ptr<SubEvent>(new SubEvent(subEventNode));
             subEvent_->setID(subEventIDRef_);
             IDManager::instance().setID(subEventIDRef_,subEvent_);
          }
          else if(subEventNode->Attribute("href")) { subEventIDRef_ = subEventNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { subEvent_ = boost::shared_ptr<SubEvent>(new SubEvent(subEventNode));}
   }

   //jointSubEventNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* jointSubEventNode = xmlNode->FirstChildElement("jointSubEvent");

   if(jointSubEventNode){jointSubEventIsNull_ = false;}
   else{jointSubEventIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- jointSubEventNode , address : " << jointSubEventNode << std::endl;
   #endif
   if(jointSubEventNode)
   {
      if(jointSubEventNode->Attribute("href") || jointSubEventNode->Attribute("id"))
      {
          if(jointSubEventNode->Attribute("id"))
          {
             jointSubEventIDRef_ = jointSubEventNode->Attribute("id");
             jointSubEvent_ = boost::shared_ptr<JointSubEvent>(new JointSubEvent(jointSubEventNode));
             jointSubEvent_->setID(jointSubEventIDRef_);
             IDManager::instance().setID(jointSubEventIDRef_,jointSubEvent_);
          }
          else if(jointSubEventNode->Attribute("href")) { jointSubEventIDRef_ = jointSubEventNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { jointSubEvent_ = boost::shared_ptr<JointSubEvent>(new JointSubEvent(jointSubEventNode));}
   }

   //jointFSubEventNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* jointFSubEventNode = xmlNode->FirstChildElement("jointFSubEvent");

   if(jointFSubEventNode){jointFSubEventIsNull_ = false;}
   else{jointFSubEventIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- jointFSubEventNode , address : " << jointFSubEventNode << std::endl;
   #endif
   if(jointFSubEventNode)
   {
      if(jointFSubEventNode->Attribute("href") || jointFSubEventNode->Attribute("id"))
      {
          if(jointFSubEventNode->Attribute("id"))
          {
             jointFSubEventIDRef_ = jointFSubEventNode->Attribute("id");
             jointFSubEvent_ = boost::shared_ptr<JointFSubEvent>(new JointFSubEvent(jointFSubEventNode));
             jointFSubEvent_->setID(jointFSubEventIDRef_);
             IDManager::instance().setID(jointFSubEventIDRef_,jointFSubEvent_);
          }
          else if(jointFSubEventNode->Attribute("href")) { jointFSubEventIDRef_ = jointFSubEventNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { jointFSubEvent_ = boost::shared_ptr<JointFSubEvent>(new JointFSubEvent(jointFSubEventNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<EventIndexTransformation> EventCheckBase::getEventIndexTransformation()
{
   if(!this->eventIndexTransformationIsNull_){
        if(eventIndexTransformationIDRef_ != ""){
             return boost::shared_static_cast<EventIndexTransformation>(IDManager::instance().getID(eventIndexTransformationIDRef_));
        }else{
             return this->eventIndexTransformation_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<EventIndexTransformation>();
   }
}
boost::shared_ptr<FunctionEvents> EventCheckBase::getFunctionEvents()
{
   if(!this->functionEventsIsNull_){
        if(functionEventsIDRef_ != ""){
             return boost::shared_static_cast<FunctionEvents>(IDManager::instance().getID(functionEventsIDRef_));
        }else{
             return this->functionEvents_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<FunctionEvents>();
   }
}
boost::shared_ptr<PayoffBase> EventCheckBase::getPayoffBase()
{
   if(!this->payoffBaseIsNull_){
        if(payoffBaseIDRef_ != ""){
             return boost::shared_static_cast<PayoffBase>(IDManager::instance().getID(payoffBaseIDRef_));
        }else{
             return this->payoffBase_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PayoffBase>();
   }
}
boost::shared_ptr<SubEvent> EventCheckBase::getSubEvent()
{
   if(!this->subEventIsNull_){
        if(subEventIDRef_ != ""){
             return boost::shared_static_cast<SubEvent>(IDManager::instance().getID(subEventIDRef_));
        }else{
             return this->subEvent_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<SubEvent>();
   }
}
boost::shared_ptr<JointSubEvent> EventCheckBase::getJointSubEvent()
{
   if(!this->jointSubEventIsNull_){
        if(jointSubEventIDRef_ != ""){
             return boost::shared_static_cast<JointSubEvent>(IDManager::instance().getID(jointSubEventIDRef_));
        }else{
             return this->jointSubEvent_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<JointSubEvent>();
   }
}
boost::shared_ptr<JointFSubEvent> EventCheckBase::getJointFSubEvent()
{
   if(!this->jointFSubEventIsNull_){
        if(jointFSubEventIDRef_ != ""){
             return boost::shared_static_cast<JointFSubEvent>(IDManager::instance().getID(jointFSubEventIDRef_));
        }else{
             return this->jointFSubEvent_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<JointFSubEvent>();
   }
}
}

